Music

shriram-sg opened the pull request:
# Music Commands Made
1. recommending music
2. searching for music

# New env variables
. SPOTIFY_CLIENT_ID
. SPOTIFY_CLIENT_SECRET


Closes #6

shriram-sg committed:
meta: add spotipy to dependencies

shriram-sg committed:
feat: fetch data with spotipy

shriram-sg committed:
chore: add async TODO + webhook test commit

shriram-sg committed:
chore: githook test

shriram-sg committed:
updating with main

shriram-sg committed:
chore: sort out conflicts in toml and lock file

There's a small chance it'll cause errors later.

shriram-sg committed:
chore: setup tekore and music client

shriram-sg committed:
feat: search for songs through spotify api

shriram-sg committed:
chore: update music command's return value to return the whole object

This makes it more robust I think. Also changed the names of the commands from "song" to "music"

shriram-sg committed:
feat: make recommend music command

shriram-sg committed:
fix: send the track's name from the api data for the recommend music command

shriram-sg committed:
chore: refactor recommend method to accomodate all three content types

shriram-sg committed:
Merge branch 'main' of https://github.com/Windsmith/utilitybot into music
chore: pulled updates from main

anand2312 commented on bot/commands/recommend.py:
```diff
@@ -35,54 +36,46 @@ def recommend_output(self, record: models.ContentRecord) -> Embed:
         async with ctx.typing():
-            data = await anime.get_anime_manga(
-                self.bot, query=name, _type=ContentType.Anime
+            if content_type == ContentType.Anime:
+                data = await anime.get_anime_manga(
+                    self.bot, query=name, _type=ContentType.Anime
+                )
+                name = data["title"]
+                url = data["siteUrl"]
+            elif content_type == ContentType.Manga:
+                data = await anime.get_anime_manga(
+                    self.bot, query=name, _type=ContentType.Manga
+                )
+                name = data["title"]
+                url = data["siteUrl"]
```
We can get away with directly passing in the `content_type` for both these cases
```suggestion
            if content_type in [ContentType.Anime, ContentType.Manga]:
                data = await anime.get_anime_manga(
                    self.bot, query=name, _type=content_type
                )
                name = data["title"]
                url = data["siteUrl"]
```

anand2312 commented on bot/commands/music.py:
```diff
@@ -0,0 +1,31 @@
+
+    @commands.command(name="music")
+    async def music(self, ctx: UtilityContext, *, name: str) -> None:
+        """Search the Spotify API (using spotipy) for a specific song."""
```
The docstring here has become outdated as we're now using tekore, but as the bot users don't need to know what lib we're using, I think we should just omit it. What do you think?
```suggestion
        """Search the Spotify API for a specific song."""
```

anand2312 commented on bot/commands/recommend.py:
```diff
@@ -4,6 +4,7 @@
 from loguru import logger
 
 from bot.backend import anime
+from bot.backend.apis import music
```
This import isn't used anywhere

anand2312 commented on bot/commands/recommend.py:
```diff
@@ -4,6 +4,7 @@
 from loguru import logger
```
This import can be removed too, looks like it's unused now

anand2312 commented on bot/backend/apis/music.py:
```diff
@@ -0,0 +1,24 @@
+    """Manage music searches and recommendations"""
+
+    def __init__(self):
+        client_id = config("SPOTIFY_CLIENT_ID")
```
Imo it's good practice to add some notes about newly added env vars to the PR description (moreover in our case, we have a .env example in the README as well that needs to be updated)

shriram-sg replied:
I'll do it soon
Update: I have done it

anand2312 commented on bot/backend/apis/music.py:
```diff
@@ -0,0 +1,24 @@
+from decouple import config
+
+
+class MusicClient:
```
The other API clients in the `backend/api` package all implement the `AbstractAPIClient` ABC (but I realize that most of the methods it enforces aren't necessary here as we're using tekore)
What's your opinion? (re: do we maintain uniformity within the `apis` package by implementing the ABC here too / just move this file out of `apis` / uniformity doesn't matter, this setup is fine)

anand2312 replied:
I think we can definitely rename the methods though (i.e `MusicClient` gets music data is already clear from it's name, so calling the method as `MusicClient.fetch_music_data` is a little redundant)

shriram-sg replied:
I think this set up is fine. The apis folder is meant for files that interact with apis (I assume), so it makes sense to keep it there even if it doesn't require the ABC.

I'll rename the method in the next commit

anand2312 commented on bot/backend/apis/music.py:
```diff
@@ -0,0 +1,24 @@
+            query, ("track",), limit=1, include_external="audio"
+        )
+
+        return tracks.items[0]
```
What's the type of this object? We should possibly add that as the return annotation for the function

anand2312 reviewed (changes requested):
Thanks for the PR, looks pretty good mostly except for a few questions

anand2312 commented on bot/backend/apis/music.py:
```diff
@@ -0,0 +1,24 @@
+        self.spotify = tekore.Spotify(app_token, asynchronous=True)
+
+    async def fetch_music_data(self, query: str):
+        """Fetch track/album/artist data in unusable form"""
```
Also, we should use the logger for every API request we make

shriram-sg committed:
Chore: Apply suggestions from code review

Co-authored-by: Anand <40204976+anand2312@users.noreply.github.com>

shriram-sg committed:
chore: implement suggested changes from pull request review

1. Implement logging and error handling
2. Change name of fetch_music_data to fetch_track_data
3. Update Readme to show the Spotify env variables
4. Remove unnecessary imports

shriram-sg committed:
fix: correct the return type of fetch_track_data

anand2312 commented on bot/backend/apis/music.py:
```diff
@@ -0,0 +1,48 @@
+                f"Could not find MUSIC with name {query} \nAPI error: nil; API is working fine"
+            )
+            raise ContentNotFoundError(f"Could not find MUSIC with name {query}")
+        except:
```
This bare `except` will silently eat up all errors; what all are we expecting here?
If you really want a catch-all here, at least let the logger know what exact error was raised via
```py
except Exception as e: 
    logger.warning(f".... API error: {e}")
    raise ContentNotFoundError(...) from e
```

anand2312 commented on bot/backend/apis/music.py:
```diff
@@ -0,0 +1,48 @@
+        client_id = config("SPOTIFY_CLIENT_ID")
+        client_secret = config("SPOTIFY_CLIENT_SECRET")
+        app_token = tekore.request_client_token(client_id, client_secret)
+        self.spotify = tekore.Spotify(app_token, asynchronous=True)
```
`tekore` provides built-in caching by using a [caching sender](https://tekore.readthedocs.io/en/stable/reference/senders.html#tekore.CachingSender) which is used like [so](https://tekore.readthedocs.io/en/stable/advanced_usage.html#using-senders)

What do you think about using it here? It might let us avoid a few of the timeouts that we've seen while testing, and caching is generally a nice idea

anand2312 commented on bot/backend/apis/music.py:
```diff
@@ -0,0 +1,48 @@
+import tekore
+from loguru import logger
+from decouple import config
+from bot.backend.exceptions import ContentNotFoundError
```
[PEP8 says that we should group imports into 3:](https://www.python.org/dev/peps/pep-0008/#imports)

- stdlib imports first
- 3rd party libs after it
- local package imports last
leaving a line between each group

```suggestion
import tekore
from decouple import config
from loguru import logger

from bot.backend.exceptions import ContentNotFoundError
```
Some formatting guides also tell us to sort imports alphabetically

anand2312 commented:
This will almost complete everything needed for #4 , except for editing/deleting from content lists

shriram-sg committed:
feat: implement caching for spotify api

anand2312 commented on bot/backend/apis/music.py:
```diff
@@ -0,0 +1,50 @@
+"""interacting with the music api"""
+
+from decouple import config
+from loguru import logger
+import tekore
```
not to be a wet blanket but `import x`s are supposed to come before the `from y import z` style imports as I'd shown in my suggested changes 😅 
```suggestion
import tekore
from decouple import config
from loguru import logger
```

anand2312 committed:
refactor: reorder imports

anand2312 reviewed (approved):
LGTM :shipit: Thanks for the contribution
